### PR TITLE
feat: send pause when hitting inflight barrier limit

### LINKED ISSUE #6682: feat: throttle source throughput to limit streaming latency
### Is your feature request related to a problem? Please describe.

As described in #6571, the barrier latency i.e. streaming end-to-end latency can increase infinitely when the system is under heavy load.

However, streaming latency is a critical metric for streaming computing, and it must be upper-bounded according to our SLA.

We need to build some negative feedback mechanisms to bound the latency.

### Describe the solution you'd like

Currently, we allow users to set the max number of in-flight barriers, but it only controls the injection of barriers. **We can enhance it by throttling the source events as well.**

> If the number of in-flight barriers reaches the upper bound i.e. 40, we may not only stop injecting barriers but also block the source executor to generate any new events. What do you guys think?
> 
> Referenced from https://github.com/risingwavelabs/risingwave/issues/6571#issuecomment-1331582075

One possible way to block new `DataChunk` events from being generated is using `Pause/Resume` barrier. Any other simpler solutions are welcomed.

### Describe alternatives you've considered

None

### Additional context

None

### PR DESCRIPTION
**Pull request by tabVersion**
Signed-off-by: tabVersion <tabvision@bupt.icu>

I hereby agree to the terms of the [Singularity Data, Inc. Contributor License Agreement](https://gist.github.com/skyzh/0663682a70b0edde7ae991492f2314cb#file-s9y_cla).

## What's changed and what's your intention?

***PLEASE DO NOT LEAVE THIS EMPTY !!!***

Please explain **IN DETAIL** what the changes are in this PR and why they are needed:

- Summarize your change (**mandatory**)
- How does this PR work? Need a brief introduction for the changed logic (optional)
- Describe clearly one logical change and avoid lazy messages (optional)
- Describe any limitations of the current code (optional)

## Checklist

- [ ] I have written necessary rustdoc comments
- [ ] I have added necessary unit tests and integration tests
- [ ] All checks passed in `./risedev check` (or alias, `./risedev c`)

## Documentation

If your pull request contains user-facing changes, please specify the types of the changes, and create a release note. Otherwise, please feel free to remove this section.

### Types of user-facing changes

Please keep the types that apply to your changes, and remove those that do not apply.

* Installation and deployment 
* Connector (sources & sinks)
* SQL commands, functions, and operators
* RisingWave cluster configuration changes
* Other (please specify in the release note below)

### Release note

Please create a release note for your changes. In the release note, focus on the impact on users, and mention the environment or conditions where the impact may occur.

## Refer to a related PR or issue link (optional)

resolve #6682

### CODE REVIEW

File: src/meta/src/barrier/schedule.rs [hunk 1]
```diff
@@ -214,39 +212,51 @@ pub struct ScheduledBarriers {
...
+                        notifiers: Default::default(),
+                        checkpoint,
+                    }
                 }
             }
+        } else {
+            Scheduled {
+                command: command_to_send.take().unwrap(),
+                notifiers: Default::default(),
+                checkpoint,
```
Comment (yezizp2012) [inline src/meta/src/barrier/schedule.rs:237]:
```suggestion
            let checkpoint = command_to_send.as_ref().unwrap().need_checkpoint() || checkpoint;
            Scheduled {
                command: command_to_send.take().unwrap(),
                notifiers: Default::default(),
                checkpoint,
```

**Comment by BugenZhao** [comment]
I suggest not reusing the `Pause` command in the meta service (while the same mutation should be acceptable). `Pause` is specialized for scaling and does much more than pausing the source executors, like:

- always do a checkpoint
- ask compute nodes to wait for epoch in `post_collect`
- pause the concurrent checkpoint

These're not throttling the source throughput wants. 

By the way, I guess making the throttle on the source executor more aggressive will solve this problem much more easily. 🤔 cc @yezizp2012

**Comment by yezizp2012** [comment]
> I suggest not reusing the `Pause` command in the meta service (while the same mutation should be acceptable). `Pause` is specialized for scaling and does much more than pausing the source executors, like:
> 
> * always do a checkpoint
> * ask compute nodes to wait for epoch in `post_collect`
> * pause the concurrent checkpoint
> 
> These're not throttling the source throughput wants.

👍 Indeed, introducing anther command with pause mutation will avoid all these unnecessary effects.

> By the way, I guess making the throttle on the source executor more aggressive will solve this problem much more easily. 🤔 cc @yezizp2012

strongly +1 as I said in #6682 🥵 , @fuyufjh What do you think?

**Comment by fuyufjh** [comment]
> > By the way, I guess making the throttle on the source executor more aggressive will solve this problem much more easily. 🤔 cc @yezizp2012
> 
> strongly +1 as I said in #6682 🥵 , @fuyufjh What do you think?

I'll be happy with any easier solution. ☺️ Can you elaborate more on "making the throttle on the source executor more aggressive"?

**Pull request closed without merging**